### DISCUSSION TITLE
BaseService refactoring

### DISCUSSION THREAD

**Discussion by adrien2p** [medusajs/medusa#1361] [RFC]
As part of the code base refactoring, here is the idea behind the baseService refactoring.

At the moment, the baseService holds the method to work with transactions and atomic phases as well as some method like buildQuery_, setMetadata_ etc.

All the services that extends the baseService are for now, implementing there own `withTransaction` method which is always the same approach as follow as explain here 👇
> Taking all the injected dependencies, set the manager with the transactionManager and set the cloned class transactionManager property.

The refactoring idea is the following:

Creating a new class `TransactionBaseService` that is abstract since there is no reason to instanciate that class as it is and it will holds the default `withTransaction` implementation as well as the `atomicPhase_` method.

The other methods from the baseService are considered as utilities method since they do not depends on the class context. Therefore, they will be move to the utility directory as utility functions.

Here is the [related PR](https://github.com/medusajs/medusa/pull/1344) and any feedback is welcome.

PS: As part of a bigger refactoring, that class will be part of the core package instead of `medusa-interfaces`

@oliverheaps @srindom here is the summary

**Comment by dwene** [comment]
That all sounds good to me!
